File: frontend/src/lib/signalling.ts
import type { RoomSnapshot, ServerMediaIceCandidate, UserProfile } from "./api";
import { apiBaseUrl } from "./api";

export type SignalPayload =
  | { type: "offer"; sdp: string }
  | { type: "answer"; sdp: string }
  | { type: "ice-candidate"; candidate: string; sdp_mid?: string; sdp_m_line_index?: number }
  | {
      type: "server-media-ice-candidate";
      candidate: string;
      sdp_mid?: string | null;
      sdp_mline_index?: number | null;
      username_fragment?: string | null;
    }
  | { type: "server-media-ice-candidates-request" }
  | { type: "server-media-ice-candidates"; candidates: ServerMediaIceCandidate[] }
  | { type: "user-joined"; user: UserProfile }
  | { type: "user-left"; user_id: string }
  | { type: "room-snapshot"; room: RoomSnapshot }
  | { type: "error"; message: string };

export type SignalMessage = {
  type: SignalPayload["type"];
  room_id: string;
  sender_id: string;
  recipient_id?: string;
  payload: SignalPayload;
};

export type PresenceState = {
  room?: RoomSnapshot;
  error?: string;
};

export function roomSocketUrl(roomId: string, userId: string, accessToken: string): string {
  const url = new URL(apiBaseUrl());
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  url.pathname = `/api/rooms/${encodeURIComponent(roomId)}/ws`;
  url.search = new URLSearchParams({ user_id: userId, access_token: accessToken }).toString();
  return url.toString();
}

export function createRoomSocket(roomId: string, userId: string, accessToken: string): WebSocket {
  return new WebSocket(roomSocketUrl(roomId, userId, accessToken));
}

export function encodeOffer(roomId: string, senderId: string, sdp: string, recipientId?: string): SignalMessage {
  return message(roomId, senderId, recipientId, { type: "offer", sdp });
}

export function encodeAnswer(roomId: string, senderId: string, sdp: string, recipientId?: string): SignalMessage {
  return message(roomId, senderId, recipientId, { type: "answer", sdp });
}

export function encodeIceCandidate(
  roomId: string,
  senderId: string,
  candidate: RTCIceCandidateInit,
  recipientId?: string
): SignalMessage {
  return message(roomId, senderId, recipientId, {
    type: "ice-candidate",
    candidate: candidate.candidate ?? "",
    sdp_mid: candidate.sdpMid ?? undefined,
    sdp_m_line_index: candidate.sdpMLineIndex ?? undefined
  });
}

export function encodeServerMediaIceCandidate(
  roomId: string,
  senderId: string,
  candidate: RTCIceCandidateInit
): SignalMessage {
  return message(roomId, senderId, senderId, {
    type: "server-media-ice-candidate",
    candidate: candidate.candidate ?? "",
    sdp_mid: candidate.sdpMid ?? null,
    sdp_mline_index: candidate.sdpMLineIndex ?? null,
    username_fragment: candidate.usernameFragment ?? null
  });
}

export function encodeServerMediaIceCandidatesRequest(roomId: string, senderId: string): SignalMessage {
  return message(roomId, senderId, senderId, { type: "server-media-ice-candidates-request" });
}

export function reducePresence(state: PresenceState, signal: SignalMessage): PresenceState {
  const payload = signal.payload;
  switch (payload.type) {
    case "room-snapshot":
      return { ...state, room: payload.room };
    case "user-joined": {
      const users = state.room?.users.filter((user) => user.id !== payload.user.id) ?? [];
      return state.room
        ? { ...state, room: { ...state.room, users: [...users, payload.user] } }
        : state;
    }
    case "user-left":
      return state.room
        ? {
            ...state,
            room: {
              ...state.room,
              users: state.room.users.filter((user) => user.id !== payload.user_id)
            }
          }
        : state;
    case "error":
      return { ...state, error: payload.message };
    case "offer":
    case "answer":
    case "ice-candidate":
    case "server-media-ice-candidate":
    case "server-media-ice-candidates-request":
    case "server-media-ice-candidates":
      return state;
  }
}

function message(
  roomId: string,
  senderId: string,
  recipientId: string | undefined,
  payload: SignalPayload
): SignalMessage {
  return {
    type: payload.type,
    room_id: roomId,
    sender_id: senderId,
    recipient_id: recipientId,
    payload
  };
}

File: crates/lyre-web/src/signalling.rs
use dashmap::DashMap;
use lyre_core::{RoomId, RoomRegistry, RoomSnapshot, UserId, UserProfile};
use lyre_webrtc::ServerMediaIceCandidate;
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SignalKind {
    Offer,
    Answer,
    IceCandidate,
    ServerMediaIceCandidate,
    ServerMediaIceCandidatesRequest,
    ServerMediaIceCandidates,
    UserJoined,
    UserLeft,
    RoomSnapshot,
    Error,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum SignalPayload {
    Offer {
        sdp: String,
    },
    Answer {
        sdp: String,
    },
    IceCandidate {
        candidate: String,
        sdp_mid: Option<String>,
        sdp_m_line_index: Option<u16>,
    },
    ServerMediaIceCandidate {
        candidate: String,
        sdp_mid: Option<String>,
        sdp_mline_index: Option<u16>,
        username_fragment: Option<String>,
    },
    ServerMediaIceCandidatesRequest,
    ServerMediaIceCandidates {
        candidates: Vec<ServerMediaIceCandidate>,
    },
    UserJoined {
        user: UserProfile,
    },
    UserLeft {
        user_id: UserId,
    },
    RoomSnapshot {
        room: RoomSnapshot,
    },
    Error {
        message: String,
    },
}

impl SignalPayload {
    fn kind(&self) -> SignalKind {
        match self {
            Self::Offer { .. } => SignalKind::Offer,
            Self::Answer { .. } => SignalKind::Answer,
            Self::IceCandidate { .. } => SignalKind::IceCandidate,
            Self::ServerMediaIceCandidate { .. } => SignalKind::ServerMediaIceCandidate,
            Self::ServerMediaIceCandidatesRequest => SignalKind::ServerMediaIceCandidatesRequest,
            Self::ServerMediaIceCandidates { .. } => SignalKind::ServerMediaIceCandidates,
            Self::UserJoined { .. } => SignalKind::UserJoined,
            Self::UserLeft { .. } => SignalKind::UserLeft,
            Self::RoomSnapshot { .. } => SignalKind::RoomSnapshot,
            Self::Error { .. } => SignalKind::Error,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SignalMessage {
    #[serde(rename = "type")]
    pub kind: SignalKind,
    pub room_id: RoomId,
    pub sender_id: UserId,
    pub recipient_id: Option<UserId>,
    pub payload: SignalPayload,
}

impl SignalMessage {
    pub fn new(
        room_id: RoomId,
        sender_id: UserId,
        recipient_id: Option<UserId>,
        payload: SignalPayload,
    ) -> Self {
        Self {
            kind: payload.kind(),
            room_id,
            sender_id,
            recipient_id,
            payload,
        }
    }

    pub fn to_self(room_id: RoomId, user_id: UserId, payload: SignalPayload) -> Self {
        Self::new(room_id, user_id.clone(), Some(user_id), payload)
    }

    pub fn error(room_id: RoomId, sender_id: UserId, message: impl Into<String>) -> Self {
        Self::new(
            room_id,
            sender_id,
            None,
            SignalPayload::Error {
                message: message.into(),
            },
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SignalRoute {
    Target(UserId),
    BroadcastExceptSender,
}

pub fn route_signal_message(
    path_room: &RoomId,
    query_user: &UserId,
    message: &SignalMessage,
) -> Result<SignalRoute, Box<SignalMessage>> {
    if &message.room_id != path_room {
        return Err(Box::new(SignalMessage::error(
            path_room.clone(),
            query_user.clone(),
            "message room_id does not match websocket room",
        )));
    }
    if &message.sender_id != query_user {
        return Err(Box::new(SignalMessage::error(
            path_room.clone(),
            query_user.clone(),
            "message sender_id does not match websocket user",
        )));
    }

    Ok(message
        .recipient_id
        .clone()
        .map(SignalRoute::Target)
        .unwrap_or(SignalRoute::BroadcastExceptSender))
}

pub type PeerSender = mpsc::UnboundedSender<SignalMessage>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SignalDelivery {
    pub delivered: usize,
}

#[derive(Debug, Default)]
pub struct PeerHub {
    peers: DashMap<RoomId, DashMap<UserId, PeerSender>>,
}

impl PeerHub {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn connect(
        &self,
        registry: &RoomRegistry,
        room_id: RoomId,
        user_id: UserId,
        tx: PeerSender,
    ) -> RoomSnapshot {
        let snapshot = registry.snapshot(room_id.clone());
        self.peers.entry(room_id).or_default().insert(user_id, tx);
        snapshot
    }

    pub fn disconnect(&self, room_id: &RoomId, user_id: &UserId) -> SignalDelivery {
        self.remove_peer(room_id, user_id);
        self.user_left(room_id, user_id)
    }

    pub fn remove_peer(&self, room_id: &RoomId, user_id: &UserId) {
        if let Some(room) = self.peers.get(room_id) {
            room.remove(user_id);
        }
    }

    pub fn user_joined(&self, room_id: &RoomId, user: UserProfile) -> SignalDelivery {
        let sender_id = user.id.clone();
        let message = SignalMessage::new(
            room_id.clone(),
            sender_id.clone(),
            None,
            SignalPayload::UserJoined { user },
        );
        self.broadcast_except(room_id, &sender_id, message)
    }

    pub fn user_left(&self, room_id: &RoomId, user_id: &UserId) -> SignalDelivery {
        let message = SignalMessage::new(
            room_id.clone(),
            user_id.clone(),
            None,
            SignalPayload::UserLeft {
                user_id: user_id.clone(),
            },
        );
        self.broadcast_except(room_id, user_id, message)
    }

    pub fn forward(&self, message: SignalMessage) -> SignalDelivery {
        if let Some(recipient_id) = message.recipient_id.clone() {
            let room_id = message.room_id.clone();
            return self.send_to(&room_id, &recipient_id, message);
        }
        let room_id = message.room_id.clone();
        let sender_id = message.sender_id.clone();
        self.broadcast_except(&room_id, &sender_id, message)
    }

    fn send_to(
        &self,
        room_id: &RoomId,
        user_id: &UserId,
        message: SignalMessage,
    ) -> SignalDelivery {
        let delivered = self
            .peers
            .get(room_id)
            .and_then(|room| room.get(user_id).map(|sender| sender.send(message).is_ok()))
            .unwrap_or(false) as usize;
        SignalDelivery { delivered }
    }

    fn broadcast_except(
        &self,
        room_id: &RoomId,
        sender_id: &UserId,
        message: SignalMessage,
    ) -> SignalDelivery {
        let delivered = self
            .peers
            .get(room_id)
            .map(|room| {
                room.iter()
                    .filter(|entry| entry.key() != sender_id)
                    .filter(|entry| entry.value().send(message.clone()).is_ok())
                    .count()
            })
            .unwrap_or(0);
        SignalDelivery { delivered }
    }
}
